linalg/ginkgo.hpp

```diff
@@ -107,169 +309,334 @@ struct ResidualLogger : gko::log::Logger
+    */
+   const std::shared_ptr<gko::LinOpFactory> GetFactory() const
+   {
+      return this->precond_gen;
+   };
+
+   /**
+    * Return a pointer to the generated preconditioner for a specific matrix
+    * (that has previously been sith with @p SetOperator).
+    */
+   const std::shared_ptr<gko::LinOp> GetGeneratedPreconditioner() const
+   {
+      return this->generated_precond;
+   };
+
+   /**
+    * Return whether this GinkgoPreconditioner object has an explicitly-
+    * generated preconditioner, built for a specific matrix.
+    */
+   bool HasGeneratedPreconditioner() const
+   {
+      return this->has_generated_precond;
+   };
+
+protected:
+   /**
+    * The Ginkgo generated solver factory object.
+    */
+   std::shared_ptr<gko::LinOpFactory> precond_gen;
+
+   /**
+    * Generated Ginkgo preconditioner for a specific matrix, created through
+    * @p SetOperator(), or a wrapped MFEM preconditioner.
+    * Must exist to use @p Mult().
+    */
+   std::shared_ptr<gko::LinOp> generated_precond;
+
+   /**
+    * The execution paradigm in Ginkgo. The choices are between
+    * `gko::OmpExecutor`, `gko::CudaExecutor` and `gko::ReferenceExecutor`
+    * and more details can be found in Ginkgo's documentation.
+    */
+   std::shared_ptr<gko::Executor> executor;
+
+   /**
+    * Whether or not we have generated a specific preconditioner for
+    * a matrix.
+    */
+   bool has_generated_precond;
+
+};
+
+/**
+* This class forms the base class for all of Ginkgo's iterative solvers.  The
+* various derived classes only take the additional data that is specific to them
+* and solve the given linear system. The entire collection of solvers that
+* Ginkgo implements is available at the Ginkgo documentation and manual pages,
+* https://ginkgo-project.github.io/ginkgo/doc/develop.
+*
+* @ingroup GinkgoWrappers
+*/
+class GinkgoIterativeSolver : public Solver
+{
+public:
+   /**
+    * Constructor.
     *
-    * ```
-    * auto omp = gko::create<gko::OmpExecutor>();
-    * using cg = gko::solver::Cg<>;
-    * auto solver_gen =
-    *     cg::build()
-    *          .with_criteria(
-    *              gko::stop::Iteration::build().with_max_iters(20u).on(omp),
-    *              gko::stop::ResidualNormReduction<>::build()
-    *                  .with_reduction_factor(1e-6)
-    *                  .on(omp))
-    *          .on(omp);
-    * auto solver = solver_gen->generate(system_matrix);
+    * The @p exec defines the paradigm where the solution is computed.
     *
-    * solver->apply(lend(rhs), lend(solution));
-    * ```
     */
-   GinkgoIterativeSolverBase(const std::string &exec_type, int print_iter,
-                             int max_num_iter, double RTOLERANCE, double ATOLERANCE);
+   GinkgoIterativeSolver(GinkgoExecutor &exec, int print_iter,
+                         int max_num_iter, double RTOLERANCE, double ATOLERANCE,
+                         bool use_implicit_res_norm);
```

Comment:
I changed the documentation in 379f0f25df8995f4cfa82bc928b81e967e1a21f9, but let me know if I need to make further changes.  It's sort of a tricky situation because the average user should never use `GinkgoIterativeSolver` (or the `use_implicit_res_norm` parameter) directly, except if making a new derived class for a Ginkgo solver that we haven't officially added to the MFEM integration yet.  This parameter is automatically set to the correct value for each derived class (`CGSolver`, `GMRESSolver`, etc.) when making the call to the base class constructor, so the user shouldn't need to worry about it.